pkg.go

```diff
@@ -27,46 +37,129 @@ type Package struct {
+	Imports []string
 }
 
 // LoadPackages loads the named packages using go list -json.
-// Unlike the go tool, an empty argument list is treated as
-// an empty list; "." must be given explicitly if desired.
-func LoadPackages(name ...string) (a []*Package, err error) {
-	if len(name) == 0 {
+// Unlike the go tool, an empty argument list is treated as an empty list; "."
+// must be given explicitly if desired.
+// IgnoredGoFiles will be processed and their dependencies resolved recursively
+// Files with a build tag of `ignore` are skipped. Files with other build tags
+// are however processed.
+func LoadPackages(names ...string) (a []*Package, err error) {
+	if len(names) == 0 {
 		return nil, nil
 	}
-	args := []string{"list", "-e", "-json"}
-	cmd := exec.Command("go", append(args, name...)...)
-	r, err := cmd.StdoutPipe()
-	if err != nil {
-		return nil, err
+	for _, i := range importPaths(names) {
+		p, err := listPackage(i)
+		if err != nil {
+			return nil, err
+		}
+		a = append(a, p)
 	}
-	cmd.Stderr = os.Stderr
-	err = cmd.Start()
+	return a, nil
+}
+
+// resolveIgnoredGoFiles for the given pkgs, recursively
+func resolveIgnoredGoFiles(pkg *Package, pc map[string]*Package) error {
+	fmt.Println("resolveIgnoredGoFiles:", pkg.ImportPath)
+	var allDeps []string
+	allDeps = append(allDeps, pkg.ImportPath)
+	allDeps = append(allDeps, pkg.Deps...)
+	allDeps = append(allDeps, pkg.TestImports...)
+	allDeps = append(allDeps, pkg.XTestImports...)
+	allDeps = uniq(allDeps)
+	spkgs, err := LoadPackages(allDeps...)
 	if err != nil {
-		return nil, err
+		return err
 	}
-	d := json.NewDecoder(r)
-	for {
-		info := new(Package)
-		err = d.Decode(info)
-		if err == io.EOF {
-			break
+	for _, sp := range spkgs {
+		if pc[sp.ImportPath] != nil {
+			continue
 		}
-		if err != nil {
-			info.Error.Err = err.Error()
+		if len(sp.IgnoredGoFiles) > 0 {
+			pc[sp.ImportPath] = sp
+			ni, nti, err := sp.ignoredGoFilesDeps()
+			fmt.Println("sp", sp.ImportPath)
+			fmt.Println("ni", ni)
+			fmt.Println("nti", nti)
+			if err != nil {
+				panic(err)
+			}
+			pkg.Deps = append(pkg.Deps, ni...)
+			pkg.TestImports = append(pkg.TestImports, nti...)
+			if len(ni) > 0 || len(nti) > 0 {
+				err := resolveIgnoredGoFiles(sp, pc)
+				if err != nil {
+					return err
+				}
+			}
 		}
-		a = append(a, info)
 	}
-	err = cmd.Wait()
-	if err != nil {
-		return nil, err
+	return nil
+}
+
+func (p *Package) ignoredGoFilesDeps() ([]string, []string, error) {
+	if p.Standard {
+		return nil, nil, nil
 	}
-	return a, nil
+
+	var buildMatch = "+build "
+	var buildFieldSplit = func(r rune) bool {
+		return unicode.IsSpace(r) || r == ','
+	}
+	var imports, testImports []string
+	for _, fname := range p.IgnoredGoFiles {
+		tgt := filepath.Join(p.Dir, fname)
+		fset := token.NewFileSet()
+		f, err := parser.ParseFile(fset, tgt, nil, parser.ParseComments)
+		if err != nil {
+			return nil, nil, err
+		}
+		if len(f.Comments) > 0 {
+			for _, c := range f.Comments {
```

Comment:
Looks like this scans all comments, but build tags have to appear near the top of the file. Might get false positives from looking at too many comments.

Also… is this necessary at all? Why not include "ignore" files too for the purposes of dependencies? Ignore doesn't necessarily mean "this file is garbage", it simply means "don't build this file into the package", same as any other unsatisfied constraint.